File: cpp_hybrid/playwright-installer.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/classified/packages/plugin-stagehand/stagehand-server/src/playwright-installer.h"

PlaywrightInstaller::PlaywrightInstaller(std::shared_ptr<Logger> logger) {
    this->logger = logger;
}

boolean PlaywrightInstaller::isPlaywrightInstalled()
{
    try
    {
        auto playwrightPath = OR((process->env->PLAYWRIGHT_BROWSERS_PATH), (join(OR((process->env->HOME), (std::string("/home/eliza"))), std::string(".cache"), std::string("ms-playwright"))));
        auto chromiumPath = join(playwrightPath, std::string("chromium-*"));
        auto hasChromium = AND((existsSync(playwrightPath)), (require(std::string("fs"))["readdirSync"](playwrightPath)["some"]([=](auto dir) mutable
        {
            return dir->startsWith(std::string("chromium-"));
        }
        )));
        return hasChromium;
    }
    catch (const any& error)
    {
        return false;
    }
}

std::shared_ptr<Promise<void>> PlaywrightInstaller::installPlaywright()
{
    this->logger->info(std::string("Installing Playwright browsers..."));
    return std::make_shared<Promise>([=](auto resolve, auto reject) mutable
    {
        auto npmPath = (process->platform == std::string("win32")) ? std::string("npm.cmd") : std::string("npm");
        auto args = array<string>{ std::string("exec"), std::string("playwright"), std::string("install"), std::string("chromium") };
        this->logger->info(std::string("Installing Playwright browsers (without system deps - should be pre-installed)"));
        auto installProcess = spawn(npmPath, args, object{
            object::pair{std::string("stdio"), std::string("pipe")}, 
            object::pair{std::string("env"), utils::assign(object{
                , 
                object::pair{std::string("DISPLAY"), OR((process->env->DISPLAY), (std::string(":99")))}
            }, process->env)}
        });
        shared stdout = string_empty;
        shared stderr = string_empty;
        installProcess->stdout->on(std::string("data"), [=](auto data) mutable
        {
            stdout += data["toString"]();
            this->logger->debug(std::string("Playwright install: ") + data["toString"]()["trim"]() + string_empty);
        }
        );
        installProcess->stderr->on(std::string("data"), [=](auto data) mutable
        {
            stderr += data["toString"]();
            this->logger->debug(std::string("Playwright install stderr: ") + data["toString"]()["trim"]() + string_empty);
        }
        );
        installProcess->on(std::string("close"), [=](auto code) mutable
        {
            if (code == 0) {
                this->logger->info(std::string("Playwright browsers installed successfully"));
                resolve();
            } else {
                auto error = std::make_shared<Error>(std::string("Playwright installation failed with code ") + code + std::string("\
stderr: ") + stderr + string_empty);
                this->logger->error(std::string("Playwright installation failed:"), error);
                reject(error);
            }
        }
        );
        installProcess->on(std::string("error"), [=](auto error) mutable
        {
            this->logger->error(std::string("Failed to start Playwright installation:"), error);
            reject(error);
        }
        );
    }
    );
}

std::shared_ptr<Promise<void>> PlaywrightInstaller::ensurePlaywrightInstalled()
{
    if (this->isPlaywrightInstalled()) {
        this->logger->info(std::string("Playwright browsers already installed"));
        return std::shared_ptr<Promise<void>>();
    }
    if (AND((this->isInstalling), (this->installPromise))) {
        this->logger->info(std::string("Playwright installation already in progress, waiting..."));
        return this->installPromise;
    }
    this->isInstalling = true;
    this->installPromise = this->installPlaywright()->finally([=]() mutable
    {
        this->isInstalling = false;
        this->installPromise = nullptr;
    }
    );
    return this->installPromise;
}

boolean PlaywrightInstaller::isReady()
{
    return OR((this->isPlaywrightInstalled()), (this->isInstalling));
}

File: classified/packages/plugin-stagehand/stagehand-server/src/playwright-installer.ts
import { execSync, spawn } from 'child_process';
import { existsSync } from 'fs';
import { join } from 'path';
import { Logger } from './logger.js';

export class PlaywrightInstaller {
  private logger: Logger;
  private isInstalling = false;
  private installPromise: Promise<void> | null = null;

  constructor(logger: Logger) {
    this.logger = logger;
  }

  /**
   * Check if Playwright browsers are installed
   */
  private isPlaywrightInstalled(): boolean {
    try {
      // Check if chromium executable exists
      const playwrightPath =
        process.env.PLAYWRIGHT_BROWSERS_PATH ||
        join(process.env.HOME || '/home/eliza', '.cache', 'ms-playwright');

      // Check for chromium directory
      const chromiumPath = join(playwrightPath, 'chromium-*');
      const hasChromium =
        existsSync(playwrightPath) &&
        require('fs')
          .readdirSync(playwrightPath)
          .some((dir: string) => dir.startsWith('chromium-'));

      return hasChromium;
    } catch (error) {
      return false;
    }
  }

  /**
   * Install Playwright browsers
   */
  private async installPlaywright(): Promise<void> {
    this.logger.info('Installing Playwright browsers...');

    return new Promise((resolve, reject) => {
      const npmPath = process.platform === 'win32' ? 'npm.cmd' : 'npm';
      const args = ['exec', 'playwright', 'install', 'chromium'];

      this.logger.info(
        'Installing Playwright browsers (without system deps - should be pre-installed)'
      );

      const installProcess = spawn(npmPath, args, {
        stdio: 'pipe',
        env: {
          ...process.env,
          DISPLAY: process.env.DISPLAY || ':99',
        },
      });

      let stdout = '';
      let stderr = '';

      installProcess.stdout.on('data', (data) => {
        stdout += data.toString();
        this.logger.debug(`Playwright install: ${data.toString().trim()}`);
      });

      installProcess.stderr.on('data', (data) => {
        stderr += data.toString();
        this.logger.debug(`Playwright install stderr: ${data.toString().trim()}`);
      });

      installProcess.on('close', (code) => {
        if (code === 0) {
          this.logger.info('Playwright browsers installed successfully');
          resolve();
        } else {
          const error = new Error(
            `Playwright installation failed with code ${code}\nstderr: ${stderr}`
          );
          this.logger.error('Playwright installation failed:', error);
          reject(error);
        }
      });

      installProcess.on('error', (error) => {
        this.logger.error('Failed to start Playwright installation:', error);
        reject(error);
      });
    });
  }

  /**
   * Ensure Playwright is installed, installing if necessary
   */
  async ensurePlaywrightInstalled(): Promise<void> {
    if (this.isPlaywrightInstalled()) {
      this.logger.info('Playwright browsers already installed');
      return;
    }

    if (this.isInstalling && this.installPromise) {
      this.logger.info('Playwright installation already in progress, waiting...');
      return this.installPromise;
    }

    this.isInstalling = true;
    this.installPromise = this.installPlaywright().finally(() => {
      this.isInstalling = false;
      this.installPromise = null;
    });

    return this.installPromise;
  }

  /**
   * Check if Playwright is ready (installed or being installed)
   */
  isReady(): boolean {
    return this.isPlaywrightInstalled() || this.isInstalling;
  }
}
